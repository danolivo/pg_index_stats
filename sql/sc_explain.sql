CREATE EXTENSION pg_index_stats;

CREATE TABLE sc_a(x integer, y text) WITH (autovacuum_enabled = off);

-- Without data no statistics created
EXPLAIN (COSTS OFF, STAT ON)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
EXPLAIN (ANALYZE, COSTS OFF, TIMING OFF, BUFFERS OFF, STAT ON, SUMMARY OFF)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
VACUUM ANALYZE sc_a;
EXPLAIN (COSTS OFF, STAT ON)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
EXPLAIN (ANALYZE, COSTS OFF, TIMING OFF, BUFFERS OFF, STAT ON, SUMMARY OFF)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';

-- Add some data and watch statistics - but only after an analyse.
INSERT INTO sc_a(x,y) (SELECT gs, 'abc'||gs%10 FROM generate_series(1,100) AS gs);
EXPLAIN (COSTS OFF, STAT ON)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
VACUUM ANALYZE sc_a;
EXPLAIN (COSTS OFF, STAT ON)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
EXPLAIN (COSTS OFF, STAT ON)
SELECT * FROM sc_a s1 JOIN sc_a s2 ON true
WHERE s1.x=1 AND s2.y LIKE 'a';

-- Check format
EXPLAIN (COSTS OFF, STAT ON, FORMAT JSON)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';

DROP EXTENSION pg_index_stats;
