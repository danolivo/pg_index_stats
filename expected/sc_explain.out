CREATE EXTENSION pg_index_stats;
CREATE TABLE sc_a(x integer, y text) WITH (autovacuum_enabled = off);
-- Without data no statistics created
EXPLAIN (COSTS OFF, STAT ON)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
                  QUERY PLAN                  
----------------------------------------------
 Seq Scan on sc_a
   Filter: ((y ~~ 'a'::text) AND (x = 1))
 Statistics:
 No statistics used during the query planning
(4 rows)

EXPLAIN (ANALYZE, COSTS OFF, TIMING OFF, BUFFERS OFF, STAT ON, SUMMARY OFF)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
                  QUERY PLAN                  
----------------------------------------------
 Seq Scan on sc_a (actual rows=0.00 loops=1)
   Filter: ((y ~~ 'a'::text) AND (x = 1))
 Statistics:
 No statistics used during the query planning
(4 rows)

VACUUM ANALYZE sc_a;
EXPLAIN (COSTS OFF, STAT ON)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
                  QUERY PLAN                  
----------------------------------------------
 Seq Scan on sc_a
   Filter: ((y ~~ 'a'::text) AND (x = 1))
 Statistics:
 No statistics used during the query planning
(4 rows)

EXPLAIN (ANALYZE, COSTS OFF, TIMING OFF, BUFFERS OFF, STAT ON, SUMMARY OFF)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
                  QUERY PLAN                  
----------------------------------------------
 Seq Scan on sc_a (actual rows=0.00 loops=1)
   Filter: ((y ~~ 'a'::text) AND (x = 1))
 Statistics:
 No statistics used during the query planning
(4 rows)

-- Add some data and watch statistics - but only after an analyse.
INSERT INTO sc_a(x,y) (SELECT gs, 'abc'||gs%10 FROM generate_series(1,100) AS gs);
EXPLAIN (COSTS OFF, STAT ON)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
                  QUERY PLAN                  
----------------------------------------------
 Seq Scan on sc_a
   Filter: ((y ~~ 'a'::text) AND (x = 1))
 Statistics:
 No statistics used during the query planning
(4 rows)

VACUUM ANALYZE sc_a;
EXPLAIN (COSTS OFF, STAT ON)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
                                                   QUERY PLAN                                                   
----------------------------------------------------------------------------------------------------------------
 Seq Scan on sc_a
   Filter: ((y ~~ 'a'::text) AND (x = 1))
 Statistics:
   sc_a.y: 1 times, stats: { MCV: 10 values, Correlation, ndistinct: 10.0000, nullfrac: 0.0000, width: 5 }
   sc_a.x: 1 times, stats: { Histogram: 0 values, Correlation, ndistinct: -1.0000, nullfrac: 0.0000, width: 4 }
(5 rows)

EXPLAIN (COSTS OFF, STAT ON)
SELECT * FROM sc_a s1 JOIN sc_a s2 ON true
WHERE s1.x=1 AND s2.y LIKE 'a';
                                                     QUERY PLAN                                                      
---------------------------------------------------------------------------------------------------------------------
 Nested Loop
   ->  Seq Scan on sc_a s1
         Filter: (x = 1)
   ->  Seq Scan on sc_a s2
         Filter: (y ~~ 'a'::text)
 Statistics:
   sc_a (s2).y: 1 times, stats: { MCV: 10 values, Correlation, ndistinct: 10.0000, nullfrac: 0.0000, width: 5 }
   sc_a (s1).x: 1 times, stats: { Histogram: 0 values, Correlation, ndistinct: -1.0000, nullfrac: 0.0000, width: 4 }
(8 rows)

-- Check format
EXPLAIN (COSTS OFF, STAT ON, FORMAT JSON)
SELECT * FROM sc_a WHERE x=1 AND y LIKE 'a';
                    QUERY PLAN                    
--------------------------------------------------
 [                                               +
   {                                             +
     "Plan": {                                   +
       "Node Type": "Seq Scan",                  +
       "Parallel Aware": false,                  +
       "Async Capable": false,                   +
       "Relation Name": "sc_a",                  +
       "Alias": "sc_a",                          +
       "Disabled": false,                        +
       "Filter": "((y ~~ 'a'::text) AND (x = 1))"+
     },                                          +
     "Statistics": {                             +
       "table": "sc_a",                          +
       "attname": "y",                           +
       "times": 1,                               +
       "stats": {                                +
         "MCV values": 10,                       +
         "Correlation": true,                    +
         "ndistinct": 10.0000,                   +
         "nullfrac": 0.0000,                     +
         "width": 5                              +
       },                                        +
       "table": "sc_a",                          +
       "attname": "x",                           +
       "times": 1,                               +
       "stats": {                                +
         "Histogram values": 0,                  +
         "Correlation": true,                    +
         "ndistinct": -1.0000,                   +
         "nullfrac": 0.0000,                     +
         "width": 4                              +
       }                                         +
     }                                           +
   }                                             +
 ]
(1 row)

DROP EXTENSION pg_index_stats;
